tutorials/dataframe/df106_HiggsToFourLeptons.py

```diff
@@ -0,0 +1,195 @@
+        } else {
+            if (abs(trackd0pv[i] / tracksigd0pv[i]) > 5 || abs(z0[i] * sin(p.Theta())) > 0.5) return false;
+        }
+    }
+    return true;
+}
+""")
+
+for s in samples:
+    # Select electron or muon trigger
+    df[s] = df[s].Filter("trigE || trigM")
+
+    # Select events with exactly four good leptons conserving charge and lepton numbers
+    # Note that all collections are RVecs and good_lep is the mask for the good leptons.
+    # The lepton types are PDG numbers and set to 11 or 13 for an electron or muon
+    # irrespective of the charge.
+    df[s] = df[s].Define("good_lep", "abs(lep_eta) < 2.5 && lep_pt > 5000 && lep_ptcone30 / lep_pt < 0.3 && lep_etcone20 / lep_pt < 0.3")\
+                 .Filter("Sum(good_lep) == 4")\
+                 .Filter("Sum(lep_charge[good_lep]) == 0")\
+                 .Define("goodlep_sumtypes", "Sum(lep_type[good_lep])")\
+                 .Filter("goodlep_sumtypes == 44 || goodlep_sumtypes == 52 || goodlep_sumtypes == 48")
+
+    # Apply additional cuts depending on lepton flavour
+    df[s] = df[s].Filter("GoodElectronsAndMuons(lep_type[good_lep], lep_pt[good_lep], lep_eta[good_lep], lep_phi[good_lep], lep_E[good_lep], lep_trackd0pvunbiased[good_lep], lep_tracksigd0pvunbiased[good_lep], lep_z0[good_lep])")
+
+    # Create new columns with the kinematics of good leptons
+    df[s] = df[s].Define("goodlep_pt", "lep_pt[good_lep]")\
+                 .Define("goodlep_eta", "lep_eta[good_lep]")\
+                 .Define("goodlep_phi", "lep_phi[good_lep]")\
+                 .Define("goodlep_E", "lep_E[good_lep]")
+
+    # Select leptons with high transverse momentum
+    df[s] = df[s].Filter("goodlep_pt[0] > 25000 && goodlep_pt[1] > 15000 && goodlep_pt[2] > 10000")
+
+# Apply luminosity, scale factors and MC weights for simulated events
+lumi = 10064.0
+for s in samples:
+    if "data" in s:
+        df[s] = df[s].Define("weight", "1.0")
+    else:
+        df[s] = df[s].Define("weight", "scaleFactor_ELE * scaleFactor_MUON * scaleFactor_LepTRIGGER * scaleFactor_PILEUP * mcWeight * {} / {} * {}".format(xsecs[s], sumws[s], lumi))
+
+# Compute invariant mass of the four lepton system and make a histogram
+ROOT.gInterpreter.Declare("""
+using Vec_t = const ROOT::RVec<float>&;
+float ComputeInvariantMass(Vec_t pt, Vec_t eta, Vec_t phi, Vec_t e)
+{
+    ROOT::Math::PtEtaPhiEVector p1(pt[0], eta[0], phi[0], e[0]);
+    ROOT::Math::PtEtaPhiEVector p2(pt[1], eta[1], phi[1], e[1]);
+    ROOT::Math::PtEtaPhiEVector p3(pt[2], eta[2], phi[2], e[2]);
+    ROOT::Math::PtEtaPhiEVector p4(pt[3], eta[3], phi[3], e[3]);
+    return (p1 + p2 + p3 + p4).M() / 1000;
+}
+""")
+
+histos = {}
+for s in samples:
+    df[s] = df[s].Define("m4l", "ComputeInvariantMass(goodlep_pt, goodlep_eta, goodlep_phi, goodlep_E)")
+    histos[s] = df[s].Histo1D(ROOT.ROOT.RDF.TH1DModel(s, "m4l", 24, 80, 170), "m4l", "weight")
```

Comment:
the double `ROOT` should be unnecessary with current master

Reply:
Thanks! I should fix this also in other places :)